### PR TITLE
ui: disassembly + thread viewer

### PR DESCRIPTION
**Pull request by 1whatleytay**
This PR adds and fixes a few things:

- Adds a new Disassembly Viewer. It can be accessed from the Debug menu. You can also jump to a segment of memory directly in the Allocation viewer.

diasm_viewer

- Adds a new individual thread viewer. It has information on the thread (its name and id), register info (pc, lr, sp, r0 - r15) and the instruction it is currently executing.

thread_viewer

- I disciplined the memory editor and told it to stop misbehaving. It will now open in the correct place 100% of the time and will never refuse to open. It has been informed that I don't like it very much.

All reviews are welcome! I made some changes to the cpu and disasm modules to make these features possible, but they might not be best practice. If something doesn't look right I'll be happy to fix it.

Tested on macOS and Windows.

### CODE REVIEW

File: src/emulator/cpu/src/cpu.cpp [hunk 1]
```diff
@@ -222,7 +218,6 @@ void stop(CPUState &state) {
 
 uint32_t read_reg(CPUState &state, size_t index) {
     assert(index >= 0);
```
Comment (VelocityRa) [inline src/emulator/cpu/src/cpu.cpp]: `index` is `size_t` (unsigned) so this check is useless too

File: src/emulator/gui/include/gui/state.h [hunk 2]
```diff
@@ -63,8 +61,15 @@ struct GuiState {
 
     DialogState common_dialog;
     GamesSelector game_selector;
-    std::unique_ptr<MemoryEditor> memory_editor;
-    size_t memory_editor_start, memory_editor_count;
+
+    MemoryEditor memory_editor;
+    size_t memory_editor_start = 0, memory_editor_count = 0;
```
Comment (VelocityRa) [inline src/emulator/gui/include/gui/state.h]: one statement per line please

File: src/emulator/gui/include/gui/state.h [hunk 3]
```diff
@@ -63,8 +61,15 @@ struct GuiState {
...
+
+    std::string disassembly_arch = "THUMB";
+    char disassembly_address[9] = "00000000", disassembly_count[5] = "100";
```
Comment (VelocityRa) [inline src/emulator/gui/include/gui/state.h]: ditto

File: src/emulator/gui/src/ui_allocations_dialog.cpp [hunk 4]
```diff
@@ -25,42 +26,50 @@ void DrawAllocationsDialog(HostState &host) {
             continue;
 
         if (ImGui::TreeNode(fmt::format("{}: {}", generation_num, generation_name).c_str())) {
-            int32_t index = -1;
-            uint32_t count = 1;
+            int32_t index = -1, count = 1;
```
Comment (VelocityRa) [inline src/emulator/gui/src/ui_allocations_dialog.cpp]: ditto

File: src/emulator/disasm/src/disasm.cpp [hunk 5]
```diff
@@ -46,15 +46,19 @@ bool init(DisasmState &state) {
     return true;
 }
 
-std::string disassemble(DisasmState &state, const uint8_t *code, size_t size, uint64_t address, bool thumb) {
+std::string disassemble(DisasmState &state, const uint8_t *code, size_t size, uint64_t address, bool thumb, uint16_t *insnSize) {
```
Comment (VelocityRa) [inline src/emulator/disasm/src/disasm.cpp]: we use macro_case, `insn_size`

**Comment by 1whatleytay** [comment]
I did a brief test on windows and it looks like it works fine (no sudden crashes or memory leaks).

File: src/emulator/gui/src/ui_disassembly_dialog.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        host.gui.disassembly.emplace_back("Nothing to disassemble.");
+        return;
+    }
+
+    uint16_t size = 1;
+    uint32_t addr = from;
+
+    for (int a = 0; a < count && size != 0; a++) {
+        // Apparently some THUMB instructions are 4 bytes long, so check all 4 just to be safe.
+        size_t addr_page = addr / KB(4), end_page = (addr + 4) / KB(4);
```
Comment (VelocityRa) [inline src/emulator/gui/src/ui_disassembly_dialog.cpp]: make separate statements

**Pull request merged**